### PR TITLE
🔒fix: CVE 2023 26115 (2)

### LINKED ISSUE #32: Regular Expression Denial of Service (ReDoS) - CVE-2023-26115
Hello,

Today DependencyTrack found [this vulnerability](https://ossindex.sonatype.org/vulnerability/CVE-2023-26115?component-type=npm&component-name=word-wrap&utm_source=dependency-track&utm_medium=integration&utm_content=v4.7.1) in my project.

The issue seems to come from [this line](https://github.com/jonschlinkert/word-wrap/blob/master/index.js#L39).

I know this project hasn't been updated for some years, so should we expect a fix? This is a transitive dependency of eslint, so I believe this issue will get a lot of attention.

Thanks!

### PR DESCRIPTION
**Pull request by OlafConijn**
Fixes https://github.com/jonschlinkert/word-wrap/issues/32

[CVE-2023-26115](https://security.snyk.io/vuln/SNYK-JS-WORDWRAP-3149973)
TLDR; ReDoS, the issue is due to Regex matching algo.

This PR proposes an alternative solution to https://github.com/jonschlinkert/word-wrap/pull/33 in which:
- `string.prototype.trimEnd` is not used, as this would break when ran on older nodejs runtimes.
- no (new) regular expressions are vulnerable to ReDoS (e.g. `/\s+$/gm`).

Performance:
- The custom `trimEnd` is slower compared to native `string.prototype.trimEnd`.
- The performance of the `trimEnd` function however is comparable to the original regular expression.

results [here](https://perf.link/#eyJpZCI6Im91aXdzenU1b3B3IiwidGl0bGUiOiJ0cmltIGVuZCIsImJlZm9yZSI6ImZ1bmN0aW9uIHRyaW1FbmQoc3RyKSB7XG4gIGxldCBsYXN0Q2hhclBvcyA9IHN0ci5sZW5ndGggLSAxO1xuICBsZXQgbGFzdENoYXIgPSBzdHJbbGFzdENoYXJQb3NdO1xuICB3aGlsZShsYXN0Q2hhciA9PT0gJyAnIHx8IGxhc3RDaGFyID09PSAnXFx0Jykge1xuICAgIGxhc3RDaGFyID0gc3RyWy0tbGFzdENoYXJQb3NdO1xuICB9XG4gIHJldHVybiBzdHIuc3Vic3RyaW5nKDAsIGxhc3RDaGFyUG9zICsgMSk7XG59IiwidGVzdHMiOlt7Im5hbWUiOiJyZWd1bGFyIGV4cHJlc3Npb24iLCJjb2RlIjoiXCJibGEgYmxhLiAgICBcIi5yZXBsYWNlKC9bIFxcdF0qJC9nKSIsInJ1bnMiOlsyMjc1MDAwLDI4NTMwMDAsMzA5OTAwMCwzMjExMDAwLDQ2NTIwMDAsMzU3OTAwMCw0NDI4MDAwLDM1OTEwMDAsMzM2MjAwMCw0NzYwMDAwLDQ5NTAwMDAsNDYzNjAwMCw0NTUwMDAwLDQ5ODAwMDAsMjAzMjAwMCw1NDQ3MDAwLDI4NTkwMDAsMTE3MDAwMCwyODQwMDAwLDQwMDQwMDAsNDQyMjAwMCwxNjIzMDAwLDM3OTAwMDAsODEyMDAwLDQ4NDMwMDAsNTc1MjAwMCwzMjQ3MDAwLDI1NzQwMDAsMjQ5MjAwMCw1MjM2MDAwLDEzOTcwMDAsNTM1MTAwMCwyMTU3MDAwLDQ4MjkwMDAsOTUwMDAsNTYzMDAwLDM0MzQwMDAsMzM1NTAwMCw1MDI5MDAwLDI0NDQwMDAsNDc2MDAwMCw3MjgwMDAsMTM3NDAwMCw0NzgwMDAsNDU5NDAwMCwxMDAwLDU0ODkwMDAsMTY3OTAwMCw4MDkwMDAsNTE2ODAwMCwyNjEzMDAwLDU1OTgwMDAsMTk5MzAwMCwzMTgwMDAsODAyMDAwLDEwMDAsMzQ0MDAwMCwyMDExMDAwLDU2NDgwMDAsNTQ5NDAwMCwzMTQ0MDAwLDcxOTAwMCw1MTgxMDAwLDQzMTAwMCw0NzEzMDAwLDExNzYwMDAsMTQyODAwMCwzMjQ5MDAwLDM2NTMwMDAsODE2MDAwLDMyNjAwMCwzMTIzMDAwLDU0MjAwMCw4ODEwMDAsODQwMDAsNDYyNzAwMCwzNDgwMDAsNTk5NzAwMCwxMDAwLDMwMzQwMDAsNDgxNjAwMCwyODg1MDAwLDU5OTAwMCwyMzM4MDAwLDUxNDAwMCw1NDcyMDAwLDE5MDAwLDQwMjAwMDAsMTgyMDAwMCwzODM2MDAwLDgwMzAwMCwxMDU1MDAwLDUyOTAwMCwzNzQ2MDAwLDgyNzAwMCwxNzE2MDAwLDQ1NzkwMDAsNDczMjAwMCwxOTYwMDAwLDE0NzMwMDBdLCJvcHMiOjI4MDkzMzB9LHsibmFtZSI6Im5hdGl2ZSB0cmltRW5kKCkiLCJjb2RlIjoiXCJibGEgYmxhLiAgICBcIi50cmltRW5kKCkiLCJydW5zIjpbNjQ0ODAwMCw3NDgwMDAwLDY0MzQwMDAsNzA2MzAwMCwxMDc1MDAwMCw3MzM1MDAwLDcxNzgwMDAsMjQ1NzAwMCw1OTY3MDAwLDk2NjMwMDAsMTI0MDMwMDAsOTQwNzAwMCwxMDQ2MzAwMCwxMjQ5NzAwMCw4ODM0MDAwLDg3ODkwMDAsMTI5MTUwMDAsMzI4NjAwMCw2MzM2MDAwLDc5NDIwMDAsODIzMTAwMCwxNTExMDAwLDQ4MzgwMDAsMzAwNTAwMCw4MDMzMDAwLDEwNTQ5MDAwLDc3NTYwMDAsNzQ4MDAwMCw2MzUyMDAwLDEyNDQwMDAwLDE4NjQwMDAsODcwMDAwMCw0MDQxMDAwLDExMzc5MDAwLDEyOTEyMDAwLDAsNTc1OTAwMCw3NDgwMDAwLDc0ODAwMDAsNTkwMDAwMCw4OTQxMDAwLDExNzkzMDAwLDM0NTkwMDAsMTAzMDUwMDAsMjUwMzAwMCwxMTMyNjAwMCw5NDg3MDAwLDg1MTAwMCw5NTIwMDAsMTI0MjMwMDAsNzExNzAwMCwxMzMzOTAwMCwxNjczMDAwLDI0OTAwMCwyNDc1MDAwLDEyMzE4MDAwLDc0ODAwMDAsNTUwMDAsOTYyNjAwMCwxMzg0MzAwMCw3ODUwMDAwLDExNDIxMDAwLDExNDA4MDAwLDEyODU1MDAwLDkzNDYwMDAsMTEwMTAwMCwxNjY4MDAwLDc0NjUwMDAsNzQ4MDAwMCwxMDk5MDAwLDkwMDAsNDc4NTAwMCwxMzU1NzAwMCwxNzc0MDAwLDEzMzY3MDAwLDEwMTQ1MDAwLDM0MTUwMDAsMTEwOTQwMDAsMTE1NDkwMDAsNzQ4MDAwMCwxMTMxOTAwMCw4MDkxMDAwLDE0OTYwMDAsNjQ5MTAwMCwxMzU0NzAwMCwxMDI3MDAwMCwxMjc4NTAwMCw5MzkwMDAsMzE3MDAwLDc0ODAwMDAsMTA2MDAwLDMyMzUwMDAsMzg1MDAwLDc3NTYwMDAsMTI0ODUwMDAsNDkxNzAwMCwxMTMzNzAwMCw4NzY2MDAwLDEzMzUyMDAwLDQ3NTgwMDBdLCJvcHMiOjcyODI3MjB9LHsibmFtZSI6ImN1c3RvbSB0cmltRW5kKCkiLCJjb2RlIjoidHJpbUVuZChcImJsYSBibGEuICAgIFwiKTsiLCJydW5zIjpbMjI2OTAwMCwxMjI4MDAwLDg3MTAwMCwxNTE5MDAwLDI1MDUwMDAsMTQxMDAwMCwxNjAwMDAwLDQ5MDMwMDAsMTYxNjAwMCwyNDk1MDAwLDQ4ODEwMDAsMjQ0MjAwMCwzOTMxMDAwLDQwNDkwMDAsMjc4NTAwMCwyNTEzMDAwLDQxNTkwMDAsMTIwNjAwMCw4NzEwMDAsNDkzNTAwMCwxMTg0MDAwLDYwNTAwMCw3NjIwMDAsMTEyNTAwMCwxNzM3MDAwLDQ4OTgwMDAsMzAwMzAwMCwyNjQ2MDAwLDc1NTAwMCwzODg5MDAwLDQ0NjYwMDAsMjE4NzAwMCw2MTgwMDAsMzQyNDAwMCwzMDU0MDAwLDg3MzAwMCw3NjMwMDAsMzE3MDAwMCwzNTEwMDAwLDI1NTIwMDAsMjA3NTAwMCwzNTMxMDAwLDg3MTAwMCwyNzg0MDAwLDQ4MDgwMDAsMjk0MDAwMCwzMDM0MDAwLDM3NzQwMDAsMzcyMDAwLDMxOTcwMDAsMjE3NzAwMCw0MDkyMDAwLDQ3MjkwMDAsNDY3NzAwMCwyMzgwMDAsNDA2ODAwMCwxNzMyMDAwLDQwMjEwMDAsMzA2OTAwMCw0MzE1MDAwLDE3MjUwMDAsMzE2OTAwMCwzMTAxMDAwLDM3MzQwMDAsMjAzMjAwMCw0MDA2MDAwLDQ5MjcwMDAsMTM1MTAwMCwxMDc4MDAwLDQzNTMwMDAsMzM3NDAwMCwxNDYyMDAwLDQ0ODcwMDAsMzMwMDAsMTAzMDAwMCwzMzg4MDAwLDg3MTAwMCwzMzAxMDAwLDQwMjEwMDAsMTM5NzAwMCwzOTQ5MDAwLDE5NDIwMDAsNDI4MjAwMCwxMjQ4MDAwLDcwMDAwLDI5ODIwMDAsMzM3OTAwMCw0NDIwMDAwLDQxMTIwMDAsMTE2NjAwMCw0MDg2MDAwLDM3MDEwMDAsNTAyMTAwMCwyNTQ4MDAwLDM2NzEwMDAsNTM4MjAwMCwzOTQwMDAwLDE1OTQwMDAsMzM5NzAwMCwxNDc0MDAwXSwib3BzIjoyNzMxMTcwfV0sInVwZGF0ZWQiOiIyMDIzLTA3LTEzVDE5OjQzOjA1Ljg2NVoifQ%3D%3D)

image

### CODE REVIEW

**Review by tlakomy — APPROVED** [review]

**Comment by sawilde** [comment]
> string.prototype.trimEnd is not used, as this would break when ran on older nodejs runtimes.

Is there a reason why we can't use the built-in `trimEnd` if it is available and only revert to a custom implementation when it is not?

**Comment by OlafConijn** [comment]
> 
We certainly could. it would be a nice performance improvement for those running node >=10

I think using the built-in `trimEnd` requires a slight bit more discussion/thought as this strips out more than just ` ` & `\t` (see [here](https://tc39.es/ecma262/multipage/text-processing.html#sec-trimstring)). As the library is used widely it is hard to oversee the effects of changing the implementation to use `trimEnd`. 

I do think it's more likely to be an improvement than a regression. I would hoever advocate for making it a different release where consumers can roll back on the `trimEnd` change without having to roll back to a version that is vulnerable to CVE 2023 26115.

**Comment by doowb** [comment]
Thanks @OlafConijn !

I'm merging and publishing this fix now.

**Comment by DonbariZ** [comment]
Download
